### PR TITLE
ARROW-1449: Implement Decimal using only Int128

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: cpp/src/arrow/compare.cc [hunk 1]
```diff
@@ -334,34 +327,9 @@ static bool IsEqualPrimitive(const PrimitiveArray& left, const PrimitiveArray& r
...
-    for (int64_t i = 0; i < left.length(); ++i) {
-      if (left.IsNull(i) != right.IsNull(i)) {
-        return false;
-      } else if (!left.IsNull(i) && (ldata[i] != rdata[i])) {
-        return false;
-      }
-    }
-    return true;
+    auto number_of_bytes_to_compare = static_cast<size_t>(byte_width * left.length());
+    return memcmp(left_data, right_data, number_of_bytes_to_compare) == 0;
```
Comment (wesm) [inline cpp/src/arrow/compare.cc]: What about nulls?

File: cpp/src/arrow/util/int128.h [hunk 2]
```diff
@@ -101,7 +102,7 @@ class ARROW_EXPORT Int128 {
   uint64_t low_bits() const { return low_bits_; }
 
   /// \brief Put the raw bytes of the value into a pointer to uint8_t.
-  void ToBytes(uint8_t** out) const;
+  Status ToBytes(std::array<uint8_t, 16>* out) const;
```
Comment (wesm) [inline cpp/src/arrow/util/int128.h]: Having not used `std::array` much I couldn't say whether this is preferable to `std::string` for transporting bytes. Seems OK for the moment, though

**Review by wesm — COMMENTED** [review]
Sad to see all this work undone, but we can always restore it later should we require decimals represented in smaller storage

**Comment by wesm** [comment]
This will need rebase after IWYU cleanups

File: cpp/src/arrow/compare.cc (hunk 1 above)
Comment (cpcloud) [reply cpp/src/arrow/compare.cc]: This is handled in the `left.null_count() > 0` case. That being `true` also means that `right.null_count() == left.null_count()`, which is checked in `BaseDataEquals`.

File: cpp/src/arrow/util/int128.h (hunk 2 above)
Comment (cpcloud) [reply cpp/src/arrow/util/int128.h]: This is similar to what I was doing before I started using `std::array`. I was using raw fixed-length C arrays, which are more cumbersome to use with the STL and just as efficient.

File: cpp/src/arrow/compare.cc (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/compare.cc]: Thanks, I misread the diff in my hurry =)

**Comment by wesm** [comment]
Looking at this again

File: cpp/src/arrow/builder.h [hunk 3]
```diff
@@ -674,6 +673,10 @@ class ARROW_EXPORT FixedSizeBinaryBuilder : public ArrayBuilder {
                          MemoryPool* pool ARROW_MEMORY_POOL_DEFAULT);
 
   Status Append(const uint8_t* value);
+
+  template <size_t NBYTES>
+  Status Append(const std::array<uint8_t, NBYTES>& data);
```
Comment (wesm) [inline cpp/src/arrow/builder.h]: So I guess if you try using this for any value of NBYTES other than 16 you'll get a linker error?

**Review by wesm — APPROVED** [review]
+1. Anything else you want to do in this patch?

**Comment by wesm** [comment]
appveyor build in progress here: https://ci.appveyor.com/project/cpcloud/arrow/build/1.0.267

File: cpp/src/arrow/builder.h (hunk 3 above)
Comment (cpcloud) [reply cpp/src/arrow/builder.h]: Good point. Let me move that implementation to the header.

**Comment by cpcloud** [comment]
I'm finished making changes here. This is ready to go on a green build.

**Comment by wesm** [comment]
:+1:

**Comment by wesm** [comment]
The Travis CI failure looks transient. Will merge as soon as the appveyor build finishes: https://ci.appveyor.com/project/cpcloud/arrow/build/1.0.272

**Pull request closed without merging**